dotnet/src/PF-Extensions/Planning.ChatStepwisePlanner/ChatStepwisePlanner.cs

```diff
@@ -0,0 +1,539 @@
+        var fewShotExamplesTemplate = fewShotExamples ?? JsonSerializer.Deserialize<FewShotExample[]>(EmbeddedResource.Read("Microsoft.SemanticKernel.Planning.ChatStepwise.Prompts.fewshotexamples.json"));
+        string promptConfigString = EmbeddedResource.Read("Microsoft.SemanticKernel.Planning.ChatStepwise.Prompts.config.json");
+        if (!string.IsNullOrEmpty(promptConfigString))
+        {
+            promptConfig = PromptTemplateConfig.FromJson(promptConfigString);
+        }
+
+        promptConfig.Completion.MaxTokens = this.Config.MaxTokens;
+
+        //this._systemStepFunction = this.ImportSemanticFunction(this._kernel, "StepwiseStep", promptTemplate, promptConfig);
+        this._nativeFunctions = this._kernel.ImportSkill(this, RestrictedSkillName);
+
+        this._context = this._kernel.CreateNewContext();
+        this._logger = this._kernel.Log;
+    }
+
+    public Plan CreatePlan(string goal)
+    {
+        if (string.IsNullOrEmpty(goal))
+        {
+            throw new PlanningException(PlanningException.ErrorCodes.InvalidGoal, "The goal specified is empty");
+        }
+
+        string functionDescriptions = this.GetFunctionDescriptions();
+
+        Plan planStep = new(this._nativeFunctions["ExecutePlan"]);
+        planStep.Parameters.Set("functionDescriptions", functionDescriptions);
+        planStep.Parameters.Set("question", goal);
+
+        //planStep.Outputs.Add("agentScratchPad");
+        planStep.Outputs.Add("stepCount");
+        planStep.Outputs.Add("skillCount");
+        planStep.Outputs.Add("stepsTaken");
+
+        Plan plan = new(goal);
+
+        plan.AddSteps(planStep);
+
+        return plan;
+    }
+
+    [SKFunction, SKName("ExecutePlan"), Description("Execute a plan")]
+    public async Task<SKContext> ExecutePlanAsync(
+        [Description("The question to answer")]
+        string question,
+        [Description("List of tool descriptions")]
+        string functionDescriptions,
+        SKContext context)
+    {
+        var stepsTaken = new List<SystemStep>();
+        this._logger?.BeginScope("ChatStepwisePlanner");
+        if (!string.IsNullOrEmpty(question))
+        {
+            this._logger?.LogInformation("Ask: {Question}", question);
+
+            // TODO MOVE
+
+            var systemPromptTemplate = EmbeddedResource.Read("Microsoft.SemanticKernel.Planning.ChatStepwise.Prompts.systemprompt.txt");
+            var fewShotExamplesTemplate = JsonSerializer.Deserialize<FewShotExample[]>(EmbeddedResource.Read("Microsoft.SemanticKernel.Planning.ChatStepwise.Prompts.fewshotexamples.json"));
+
+            // TODO use semantic function to fill in function descriptions
+            string currentTime = DateTimeOffset.UtcNow.ToString("f", DateTimeFormatInfo.InvariantInfo);
+            ChatHistory history = this._chatCompletion.CreateNewChat(systemPromptTemplate.Replace("{{$functionDescriptions}}", functionDescriptions).Replace("{{time.UtcNow}}", currentTime));
```

Comment:
>{{$functionDescriptions}}", functionDescriptions).Replace("{{time.UtcNow}} [](http://example.com/codeflow?start=99&length=74)

{{$foo}} (variable assignment) and {time.UtcNow} (native function call) are SK syntax which normally assigned by the kernel.
What was the thought of string.replace them ahead of time?

Reply:
I had to change the method so it was no longer a semantic function in order to use the chat completion. As a result, we don't have the SK replacement in the prompts.

This was just a hack to get it to work without being a function. I am planning to discuss a better implementation as part of integrating this directly into SK.

Reply:
No, I imagine we could still support time when generating the system prompt, it is a side effect of calling the LLM using lower level APIs:

var llmResponse = (await this._chatCompletion.GetChatCompletionsAsync(history).ConfigureAwait(false));

rather than the way it is done in StepwisePlanner:

 this._systemStepFunction = this.ImportSemanticFunction(this._kernel, "StepwiseStep", promptTemplate, promptConfig);
"